src/components/Reown/NanoContract/BaseNanoContractRequest.js

```diff
@@ -0,0 +1,528 @@
+  const notRegistered = useMemo(() => (
+    notInitialize && registeredNc == null
+  ), [notInitialize, registeredNc]);
+  const showRequest = !notRegistered;
+
+  // Check if we have enough balance for deposit actions
+  const hasInsufficientBalance = useMemo(() => {
+    if (!checkInsufficientBalance || !nano.actions) return false;
+
+    // Track running balance per token
+    const tokenBalances = {};
+
+    for (const { type, token, amount } of nano.actions) {
+      if (type === 'deposit') {
+        // Initialize token balance if first time seeing this token
+        if (!(token in tokenBalances)) {
+          const balance = tokensBalance[token]?.data?.available || 0n;
+          tokenBalances[token] = balance;
+        }
+
+        tokenBalances[token] -= amount;
+
+        if (tokenBalances[token] < 0n) {
+          // Early exit if we go negative
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }, [checkInsufficientBalance, nano.actions, tokensBalance]);
+
+  // Request first address and addresses for selection when component mounts
+  useEffect(() => {
+    if (!firstAddress.address && !firstAddress.error) {
+      dispatch(firstAddressRequest());
+    }
+    dispatch(selectAddressAddressesRequest());
+  }, []);
+
+  // Update ncAddress when firstAddress becomes available
+  useEffect(() => {
+    if (firstAddress.address && !ncAddress) {
+      setNcAddress(firstAddress.address);
+    }
+
+    if (notRegistered && !isNcRegistering && !firstAddress.address && !firstAddress.error) {
+      dispatch(firstAddressRequest());
+    }
+  }, [firstAddress, notRegistered, isNcRegistering]);
+
+  // When nano contract is registered it should set the caller address
+  useEffect(() => {
+    if (!ncAddress && registeredNc?.address) {
+      setNcAddress(registeredNc.address);
+    }
+  }, [registeredNc, ncAddress]);
+
+  // Handle successful transaction navigation
+  useEffect(() => {
+    if (status === statusConfig.statusConstants.SUCCESSFUL) {
+      navigation.navigate(
+        'SuccessFeedbackScreen',
+        {
+          title: t`Success!`,
+          message: t`Transaction successfully sent.`,
+        }
+      );
+      dispatch(statusConfig.setReadyAction());
+    }
+  }, [status]);
+
+  // Request blueprint info and token data when component mounts
+  useEffect(() => {
+    // This component should always start with ready state
+    dispatch(statusConfig.setReadyAction());
+
+    // Do nothing if nano contract is not registered and don't call initialize method
+    if (notRegistered) return undefined;
+
+    if (!blueprintInfo) {
+      dispatch(nanoContractBlueprintInfoRequest(nano.blueprintId));
+    }
+
+    // Request token data for each unknown token present in actions
+    const unknownTokensUid = [];
+    const actionTokensUid = nano.actions?.map((each) => each.token) || [];
+    actionTokensUid.forEach((uid) => {
+      if (uid !== DEFAULT_TOKEN.uid && !(uid in knownTokens)) {
```

Comment:
question: Also here, in case the Default token is not the Native token, the native token may be added to the unknown tokens list. This is probably not desirable.